### ISSUE TITLE
reproduce ave dataset using vgg-19 features

### ISSUE THREAD

**Issue by yuhanying** [JacobChalk/TIM#41]
Hi, thank you for sharing the code. I’m working on a project using your implementation and have a few questions regarding some aspects I encountered. I’d really appreciate your help!

1. I’m using the features from [(https://github.com/YapengTian/AVE-ECCV18)]

![Image](https://github.com/user-attachments/assets/3663ef6a-ff12-4f0c-9be4-5f5d9fc84f66)
I also used additional metadata files and found that segments from the same video can have two different labels.
In the sliding_window.py, I saw the following settings:

![Image](https://github.com/user-attachments/assets/ef93df5c-cc08-4a61-9912-ab8af1a2b8ee)
Am I correct in understanding that these values indicate the number of labels per video segment?

2. During training, I used the validation set to monitor performance and avoid overfitting, and I saved the best model based on validation results (as shown in the first figure).
Then I used this best model to evaluate on the test set (as shown in the second figure), but the test accuracy was very poor.
Is there anything specific I should be aware of when running inference on the test set? I want to make sure I'm not missing any required preprocessing or setup steps.

![Image](https://github.com/user-attachments/assets/fb060d3a-a43d-4884-8c10-038bec3a516c)

![Image](https://github.com/user-attachments/assets/8b4e042b-2dd0-4bb5-be22-bcdc13493add)

3.In the recognition TIM.py part of the code, I noticed that the mask is only applied up to self.num_feats. However, since the input includes both visual and audio features (with each occupying num_feats), shouldn’t the mask be applied up to 2 * self.num_feats?

![Image](https://github.com/user-attachments/assets/b0141f21-7b79-4324-8d3a-5dbd0ec4fa83)

Thank you very much for your time and for open-sourcing your work. Looking forward to your response!

**Comment by JacobChalk** [comment]
Hi,

Thank you for using the TIM code! Hopefully we can clarify, to answer your points:

1. The variables `self.v_num_aug = 1` and `self.a_num_aug = 1` refer to the number of augmented feature sets that have been extracted, rather than the number of labels. Their use can be seen in `time_interval_machine.datasets.sliding_window` on [L358](https://github.com/JacobChalk/TIM/blob/c7cb2935fb6736db3e4328e749ad25a2433a1654/recognition/time_interval_machine/datasets/sliding_window.py#L358), meaning it assumes the features are of shape [N, num_aug, ...]. This doesn't apply as much to the ECCV AVE features because we didn't augment them, but you still need to unsqueeze dim 1 so your features are of shape `(4143, 1, 10, 7, 7, 513)` and `(4143, 1, 10, 128)` otherwise I think it will randomly sample out of the 10 dim which is unintended.
2. When running inference on the test set, you should have an almost identical set-up to the validation, with the exception of passing in dummy labels of -1 as the ground truth is unknown. This could potentially be connected to issue 1 if the features are not of the correct shape. I would double check that extra dim of 1 is there in the data loader!
3. You are 100% correct in thinking this. There is a slightly hidden line that doubles the number of features if the model is audio-visual on [L87](https://github.com/JacobChalk/TIM/blob/c7cb2935fb6736db3e4328e749ad25a2433a1654/recognition/time_interval_machine/models/tim.py#L87) in `time_interval_machine.models.tim`. This means the mask is actually being applied up to 2*num_feats if the audio_visual model is used.

By unsqueezing the features in dim 1, you should see improved results both on the val and test set! I'd recommend retraining the model so it applies the AVGA pooling as intended. If this doesn't work, it may be some other issue in the data pre-processing stage which we can then look further into. Hopefully this clears up any ambiguities and hope this helps!

**Comment by yuhanying** [comment]
Also, I have a follow-up question regarding point 1.
I noticed that some video segments (e.g., from 0–10 seconds) are annotated with two different labels in the metadata.
![Image](https://github.com/user-attachments/assets/10cd01f6-c00a-432d-9a3e-c066408913b6)
How did you handle this case during training?

For example, in one case, a video segment was annotated with both labels. My current approach is to reshape the features to something like (num_aug, 10, num_labels, ...):

![Image](https://github.com/user-attachments/assets/d3e32311-d978-4317-9c14-a4b7d4d47856)

![Image](https://github.com/user-attachments/assets/f59ff6d2-361c-48cd-a259-18face8e19c4)
Is this an appropriate way to represent multiple labels for a segment? Or is there a better/recommended way to deal with this scenario in your implementation?

Thanks again for your time and help!

**Comment by JacobChalk** [comment]
This will be handled through the dataset in training. Specifically, for each window, [L228](https://github.com/JacobChalk/TIM/blob/c7cb2935fb6736db3e4328e749ad25a2433a1654/recognition/time_interval_machine/datasets/sliding_window.py#L228) all the queries will be remembered and used to prompt TIM for classification. This means that for that specific video there will be 20 queries, 10 for class 18 and 10 for class 28. These cases aren't explicitly handled in training as TIM was designed mainly around the EPIC-KITCHENS/Perception Test dataset where multi-label classification like this rarely occurs, so it will likely only get 1 correct out of both of these.

In summary, all of these actions/labels will still be seen by creating multiple queries for the same timestamps and assigning different labels. However, this goes slightly against the assumption TIM has (that the exact same query will have exclusively one action) so will not have optimal results in these cases.
